044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,156 @@
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is currently hard-coded to 9090.
+If it has a broker process it should be listening on the replication port which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
+* The current-state metric could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
+* The broker nodes will not move to a RUNNING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: controller is listening on 9090
+* Readiness: it is possible to make a TCP connection to port 9090 on the controller
+(i.e. client socket gets into the `ESTABLISHED` TCP socket state)
```

Comment:
Since there is a gap between the process starting to listen on that port and actually accepting connections, the intention was that the readiness check does involve actually trying to connect. For the liveness check I guess we would stick with the existing test we've used elsewhere of:
```
netstat -lnt | grep -Eq 'tcp6?[[:space:]]+[0-9]+[[:space:]]+[0-9]+[[:space:]]+[^ ]+:9090.*LISTEN[[:space:]]*'
```

Then for the readiness we could use something like netcat couldn't we? E.g. `nc -w 5 -v localhost 9090 </dev/null; echo $?`

Reply:
I suspect this would create terrible mess in the broker logs ... there will be errors about TLS handshake etc. So I would be very careful about doing any `nc` stuff to it.

Reply:
I hadn't considered the impact that might have on the controller logs, that could lead users to think they have a misbehaving client, when the actually don't. Looking at https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ControllerServer.scala there seems to be fewer actions taking place between the controller starting to listen and accepting connections, so perhaps there is minimal benefit to having separate checks. @tombentley what do you think? Would it be reasonable to have the same liveness and readiness checks for the controller? Specifically have the check be that the controller is listening on the port.